### ISSUE TITLE
Suggestion about testing streamlink with real URLs

### ISSUE THREAD

**Issue by amurzeau** [streamlink/streamlink#1469]
### Checklist

- [ ] This is a bug report.
- [ ] This is a feature request.
- [x] This is a plugin (improvement) request.
- [x] I have read the contribution guidelines.

### Description

As streaming platforms tend to change over time and break Streamlink, what do you think about adding a test of all plugins against real website.
The goal is:
- to detect as early as possible breaking changes in streaming platforms that make plugins to fail
- have an overview of all plugins state (some might not be used by anyone or for a now closed platform)

This test would do:
- For each input URL:
  - Run plugin detection on it which choose the right plugin to handle the URL and retrieve metadata (the python stream class)
  - If no error is raised, the test pass, else it fails for that plugin, indicating that the plugin is broken

Things to consider:
- This require to have real sample URLs for all platforms we want to test
  - Some platforms might always require a channel which must be online at the time of the test. a 24/24 channel can be sufficient.
- Having working URLs somewhere in Streamlink code base might be against policies (although many unit tests already have real URLs)
- This new test need internet to work. But it's a integration test and not a unit test.
 - This test can be implemented as a separate script outside regular python unit tests so devs only run this one if they want to.

What do you think about it ?

[Love Streamlink? Please consider supporting our collective. Thanks!](https://opencollective.com/streamlink/donate)

**Comment by bastimeyer** [comment]
While I agree that there should be *real* plugin tests for making sure that they actually work, I think that there's a lot which can prevent the tests to pass. As you already said, running integration tests on actual URLs requires certain channels to be accessible all the time or channels need to be chosen dynamically beforehand by making API calls or by parsing the website listing available channels. This alone can and will fail at some point, which makes implementing those tests rather difficult. The streaming service could also block traffic from the CI tool, making the tests useless. I agree that these tests should be run so that they don't let a build fail, as an optional step. Maybe the tests could be run daily once and update a column on the plugins list on the website with the current status of the plguin (and its test).

**Comment by amurzeau** [comment]
With travis, it is possible to run periodic (cron) builds.
Plugins that cannot be tested (because of 24/24 channel availability) could be ignored to avoid maintenance burden on testing them.
I will try to see if it is possible technically to do that kind of tests.

**Comment by back-to** [comment]
travis is probably a bad idea, but **locally** it could work with a random and timedelay script

you could write a _website crawler_ for none 24/7 channels,
but some plugins won't work because of geo-blocking / account-needed ...

the probleme is probably to make the output look good

**Comment by gravyboat** [comment]
I'm with @bastimeyer and @back-to. Having real tests run in any automated fashion would be a huge problem for people contributing if their PR is broken because of one plugin or geo-location or account issue. There is also the issue of a whack-a-mole style cropping up where we're constantly fixing plugins that people might not even be using, and that's a terrible use of our very limited time not to mention very annoying. At this point I think we simply support too many plugins that cover too many weird cases to make this a worthy use of our time to detect and deal with automatically. This seems like a really big time investment for a small win (someone not having to make an issue for a very specific plugin).

**Comment by amurzeau** [comment]
Yes I agree on the time consuming task this can be, if a plugin is broken and widely used, someone will report it being broken.
I'm now closing this :)

**Comment by gravyboat** [comment]
@amurzeau Okay, maybe we can bring it back up if we end up reducing the number of plugins we support in the future? I definitely agree with you that it's a good idea and I don't mean to shoot it down at all.

**Comment by amurzeau** [comment]
It was just a wild suggestion so don't worry :p. Maybe this can also help tracking broken plugins to remove them instead.
But it is somewhat a good challenge to keep good URLs, given multiple possible constraints.
I made the scripts when I opened this issue, to see the amount of work required. Both are fairly small as is but everything will be in details.

Also about crawling, this would be a good amount of work I think, to have a good working server replicating original one.
